### PR TITLE
QgsMapLayerAction: add legend context menu entry for action on selection

### PR DESCRIPTION
**Pull request by 3nids**
A new menu entry (Actions on selection) is added to the layer tree context menu
It is shown only when required (layer has selection and actions can be run on the given layer)

image

### CODE REVIEW

**Comment by nirvn** [comment]
Why are the actions seemingly duplicated in the screenshot above (a pair of action just below rename, as well as the sub menu)?

**Comment by 3nids** [comment]
@nirvn good catch ;) they currently comes from a plugin which creates both the map layer action + add entries in the context menu. 

if community agrees on this addition, I will also add the `Layer` target to the menu.

I was thinking about adding some "context" flag in the map layer action (legend context menu, attribute table, identify menu, etc.) to be able to customize where the actions will be shown.

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 1]
```diff
@@ -272,6 +273,34 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
...
+
+        const QList<QgsMapLayerAction *> constRegisteredActions = QgsGui::mapLayerActionRegistry()->mapLayerActions( vlayer, target );
+        if ( !constRegisteredActions.isEmpty() )
+        {
+          QMenu *actionMenu = menu->addMenu( tr( "Actions on selection (%1)" ).arg( selectionCount ) );
+          for ( QgsMapLayerAction *action : constRegisteredActions )
+          {
+            if ( target == QgsMapLayerAction::Target::SingleFeature )
+            {
+              actionMenu->addAction( action->text(), action, [ = ]() {action->triggerForFeature( vlayer,  new QgsFeature( vlayer->selectedFeatures().at( 0 ) ) );} );
```
Comment (3nids) [inline src/app/qgsapplayertreeviewmenuprovider.cpp]: I believe that this is leaking.
How can I handle this?

Comment (nyalldawson) [reply src/app/qgsapplayertreeviewmenuprovider.cpp]: Change the api -- it shouldn't be a pointer here

Comment (3nids) [reply src/app/qgsapplayertreeviewmenuprovider.cpp]: @nyalldawson done, works for you?

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 2]
```diff
@@ -272,6 +273,34 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
...
+
+        const QList<QgsMapLayerAction *> constRegisteredActions = QgsGui::mapLayerActionRegistry()->mapLayerActions( vlayer, target );
+        if ( !constRegisteredActions.isEmpty() )
+        {
+          QMenu *actionMenu = menu->addMenu( tr( "Actions on selection (%1)" ).arg( selectionCount ) );
+          for ( QgsMapLayerAction *action : constRegisteredActions )
+          {
+            if ( target == QgsMapLayerAction::Target::SingleFeature )
+            {
+              actionMenu->addAction( action->text(), action, [ = ]() {action->triggerForFeature( vlayer,  QgsFeature( vlayer->selectedFeatures().at( 0 ) ) );} );
```
Comment (nyalldawson) [inline src/app/qgsapplayertreeviewmenuprovider.cpp]:
```suggestion
              actionMenu->addAction( action->text(), action, [ = ]() { action->triggerForFeature( vlayer,  vlayer->selectedFeatures().at( 0 ) ); } );
```

File: src/app/qgsidentifyresultsdialog.cpp [hunk 3]
```diff
@@ -1479,7 +1479,7 @@ void QgsIdentifyResultsDialog::doMapLayerAction( QTreeWidgetItem *item, QgsMapLa
     return;
 
   int featIdx = featItem->data( 0, Qt::UserRole + 1 ).toInt();
-  action->triggerForFeature( layer, &mFeatures[ featIdx ] );
+  action->triggerForFeature( layer, QgsFeature( mFeatures[ featIdx ] ) );
```
Comment (nyalldawson) [inline src/app/qgsidentifyresultsdialog.cpp]:
```suggestion
  action->triggerForFeature( layer, mFeatures[ featIdx ] );
```

File: src/app/qgsidentifyresultsdialog.cpp [hunk 4]
```diff
@@ -2186,7 +2186,7 @@ void QgsIdentifyResultsDialog::formatChanged( int index )
 
 void QgsIdentifyResultsDialogMapLayerAction::execute()
 {
-  mAction->triggerForFeature( mLayer, mFeature );
+  mAction->triggerForFeature( mLayer, QgsFeature( *mFeature ) );
```
Comment (nyalldawson) [inline src/app/qgsidentifyresultsdialog.cpp]:
```suggestion
  mAction->triggerForFeature( mLayer, *mFeature );
```

File: src/gui/attributetable/qgsattributetableview.cpp [hunk 5]
```diff
@@ -449,7 +449,7 @@ void QgsAttributeTableView::actionTriggered()
     QgsMapLayerAction *layerAction = qobject_cast<QgsMapLayerAction *>( object );
     if ( layerAction )
     {
-      layerAction->triggerForFeature( mFilterModel->layer(), &f );
+      layerAction->triggerForFeature( mFilterModel->layer(), QgsFeature( f ) );
```
Comment (nyalldawson) [inline src/gui/attributetable/qgsattributetableview.cpp]:
```suggestion
      layerAction->triggerForFeature( mFilterModel->layer(), f );
```

File: src/gui/qgsidentifymenu.cpp [hunk 6]
```diff
@@ -498,7 +498,7 @@ void QgsIdentifyMenu::triggerMapLayerAction()
       {
         if ( result.mFeature.id() == actData.mFeatureId )
         {
-          actData.mMapLayerAction->triggerForFeature( actData.mLayer, new QgsFeature( result.mFeature ) );
+          actData.mMapLayerAction->triggerForFeature( actData.mLayer, QgsFeature( result.mFeature ) );
```
Comment (nyalldawson) [inline src/gui/qgsidentifymenu.cpp]:
```suggestion
          actData.mMapLayerAction->triggerForFeature( actData.mLayer, result.mFeature );
```

Comment (nyalldawson) [reply src/gui/qgsidentifymenu.cpp]: Looks like you fixed another leak here...

**Comment by nyalldawson** [comment]
Yep, much nicer

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 7]
```diff
@@ -272,6 +273,34 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
...
+        QgsMapLayerAction::Target target;
+        if ( selectionCount == 1 )
+          target = QgsMapLayerAction::Target::SingleFeature;
+        else
+          target = QgsMapLayerAction::Target::MultipleFeatures;
+
+        const QList<QgsMapLayerAction *> constRegisteredActions = QgsGui::mapLayerActionRegistry()->mapLayerActions( vlayer, target );
+        if ( !constRegisteredActions.isEmpty() )
+        {
+          QMenu *actionMenu = menu->addMenu( tr( "Actions on selection (%1)" ).arg( selectionCount ) );
```
Comment (nyalldawson) [inline src/app/qgsapplayertreeviewmenuprovider.cpp]:
```suggestion
          QMenu *actionMenu = menu->addMenu( tr( "Actions on Selection (%1)" ).arg( selectionCount ) );
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
LGTM

**Pull request merged**